From 9b9d9330710585fe1916c9cb84e4f82663c8d33b Mon Sep 17 00:00:00 2001
From: darkyzhou <me@zqy.io>
Date: Mon, 17 Nov 2025 22:59:02 +0800
Subject: [PATCH 06/21] AOSCOS: build: correct Clang builtins path

With AOSC OS, LLVM is installed in a versioned prefix: /usr/lib/llvm-$VER.

Correct the builtins path accordingly.

Link: https://github.com/AOSC-Dev/chromium-loongarch64/blob/chromium-142.0.7444.134-1/chromium/chromium-142.0.7444.134.3002-fix-clang-builtins-path.diff
---
 build/config/clang/BUILD.gn  | 57 +++++++++++++++++++++++++-----------
 build/config/clang/clang.gni |  4 +++
 2 files changed, 44 insertions(+), 17 deletions(-)

diff --git a/build/config/clang/BUILD.gn b/build/config/clang/BUILD.gn
index e05032c..a5f6239 100644
--- a/build/config/clang/BUILD.gn
+++ b/build/config/clang/BUILD.gn
@@ -185,24 +185,47 @@ template("clang_lib") {
       } else if (is_apple) {
         _dir = "darwin"
       } else if (is_linux || is_chromeos) {
-        if (current_cpu == "x64") {
-          _dir = "x86_64-unknown-linux-gnu"
-        } else if (current_cpu == "x86") {
-          _dir = "i386-unknown-linux-gnu"
-        } else if (current_cpu == "arm") {
-          _dir = "armv7-unknown-linux-gnueabihf"
-        } else if (current_cpu == "arm64") {
-          _dir = "aarch64-unknown-linux-gnu"
-        } else if (current_cpu == "loong64") {
-          _dir = "loongarch64-unknown-linux-gnu"
-        } else if (current_cpu == "riscv64") {
-          _dir = "riscv64-unknown-linux-gnu"
-        } else if (current_cpu == "ppc64") {
-          _dir = "ppc64le-unknown-linux-gnu"
-        } else if (current_cpu == "s390x") {
-          _dir = "s390x-unknown-linux-gnu"
+        if (clang_use_per_target_runtime_dir) {
+          if (current_cpu == "x64") {
+            _dir = "x86_64-unknown-linux-gnu"
+          } else if (current_cpu == "x86") {
+            _dir = "i386-unknown-linux-gnu"
+          } else if (current_cpu == "arm") {
+            _dir = "armv7-unknown-linux-gnueabihf"
+          } else if (current_cpu == "arm64") {
+            _dir = "aarch64-unknown-linux-gnu"
+          } else if (current_cpu == "loong64") {
+            _dir = "loongarch64-unknown-linux-gnu"
+          } else if (current_cpu == "riscv64") {
+            _dir = "riscv64-unknown-linux-gnu"
+          } else if (current_cpu == "ppc64") {
+            _dir = "ppc64le-unknown-linux-gnu"
+          } else if (current_cpu == "s390x") {
+            _dir = "s390x-unknown-linux-gnu"
+          } else {
+            assert(false)  # Unhandled cpu type
+          }
         } else {
-          assert(false)  # Unhandled cpu type
+          _dir = "linux"
+          if (current_cpu == "x64") {
+            _suffix = "-x86_64"
+          } else if (current_cpu == "x86") {
+            _suffix = "-i386"
+          } else if (current_cpu == "arm") {
+            _suffix = "-armv7"
+          } else if (current_cpu == "arm64") {
+            _suffix = "-aarch64"
+          } else if (current_cpu == "loong64") {
+            _suffix = "-loongarch64"
+          } else if (current_cpu == "riscv64") {
+            _suffix = "-riscv64"
+          } else if (current_cpu == "ppc64") {
+            _suffix = "-ppc64le"
+          } else if (current_cpu == "s390x") {
+            _suffix = "-s390x"
+          } else {
+            assert(false)  # Unhandled cpu type
+          }
         }
       } else if (is_fuchsia) {
         if (current_cpu == "x64") {
diff --git a/build/config/clang/clang.gni b/build/config/clang/clang.gni
index e49005d..5bc460e 100644
--- a/build/config/clang/clang.gni
+++ b/build/config/clang/clang.gni
@@ -77,6 +77,10 @@ declare_args() {
 
   clang_base_path = default_clang_base_path
 
+  # Set to false for distro LLVM packages that install compiler-rt archives
+  # as lib/<os>/libclang_rt.<name>-<arch>.a instead of per-target dirs.
+  clang_use_per_target_runtime_dir = true
+
   # Specifies whether or not bitcode should be embedded during compilation.
   # This is used for creating a MLGO corpus from Chromium in the non-ThinLTO case.
   clang_embed_bitcode = false
-- 
2.54.0

